pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -650,22 +656,41 @@ impl RemoteTimelineClient {
         // from latest_files, but not yet scheduled for deletion. Use a closure
         // to syntactically forbid ? or bail! calls here.
         let no_bail_here = || {
-            for name in names {
-                if upload_queue.latest_files.remove(name).is_some() {
-                    upload_queue.latest_files_changes_since_metadata_upload_scheduled += 1;
-                }
-            }
+            // Decorate our list of names with each name's generation, dropping
+            // makes that are unexpectedly missing from our metadata.
+            let with_generations: Vec<_> = names
+                .iter()
+                .filter_map(|name| {
+                    // Remove from latest_files, learning the file's remote generation in the process
+                    let meta = upload_queue.latest_files.remove(name);
+
+                    if let Some(meta) = meta {
+                        upload_queue.latest_files_changes_since_metadata_upload_scheduled += 1;
+                        Some((name, meta.generation))
```

Comment:
The outcome of this discussion will affect
* https://github.com/neondatabase/neon/pull/5140#discussion_r1309988830 
* https://github.com/neondatabase/neon/pull/5140#discussion_r1309993602

Reply:
Discussed on slack -- the metadata carries the generation because that's the authoritative location in the index, after we remove `timeline_layers` (https://github.com/neondatabase/neon/pull/5153).